lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/person/accountedit/PersonAccountEditScreen.kt

```diff
@@ -52,19 +57,22 @@ fun PersonAccountEditScreen(
 
         if (uiState.usernameVisible){
             OutlinedTextField(
-                modifier = Modifier.fillMaxWidth().testTag("username"),
+                modifier = Modifier.fillMaxWidth().testTag("username").onKeyEvent { keyEvent ->
+                    if (keyEvent.type == KeyEventType.KeyDown) {
+                        FilterUsernameUseCase.shouldBlockKeyEvent(keyEvent.utf16CodePoint.toChar())
+                    } else false
+                },
                 value = uiState.personAccount?.username ?: "",
                 maxLines = 1,
                 label = {
                     Text(stringResource(MR.strings.username) + "*")
                 },
-                onValueChange = {
-                    onChange(
-                        uiState.personAccount?.copy(
-                            username = it
-                        )
-                    )
-
+                onValueChange = { newValue ->
+                    newValue.lastOrNull()?.let { lastChar ->
```

Comment:
I already pointed out on [previous pull request](https://github.com/UstadMobile/UstadMobile/pull/986) - _"This logic fails when anytime the user has changed any character which is not at the end of the text field."_ - comment on LoginViewModel.kt

Please remove all value change event processing based on the flawed assumption that the last character in the string is the one that was just changed.

Reply:
Use the filter username use case on change. If the username is valid - filter username use case won't change it. If something invalid has been entered via IME (or copy/paste) - then this will catch it.